src/ImageSharp/Formats/Jpeg/Components/Decoder/HuffmanScanDecoder.cs

```diff
@@ -704,355 +568,302 @@ private unsafe void ParseProgressiveDataNonInterleaved()
+            buffer.Remain -= size;
+
+            return symbol;
         }
 
-        [MethodImpl(InliningOptions.ColdPath)]
-        private unsafe int DecodeHuffmanSlow(ref HuffmanTable table)
+        [MethodImpl(InliningOptions.ShortMethod)]
+        private void Reset()
         {
-            // Naive test is to shift the code_buffer down so k bits are
-            // valid, then test against MaxCode. To speed this up, we've
-            // preshifted maxcode left so that it has (16-k) 0s at the
-            // end; in other words, regardless of the number of bits, it
-            // wants to be compared against something shifted to have 16;
-            // that way we don't need to shift inside the loop.
-            uint temp = this.codeBuffer >> 16;
-            int k;
-            for (k = FastBits + 1; ; ++k)
-            {
-                if (temp < table.MaxCode[k])
-                {
-                    break;
-                }
-            }
-
-            if (k == 17)
-            {
-                // Error! code not found
-                this.codeBits -= 16;
-                return -1;
-            }
-
-            if (k > this.codeBits)
+            for (int i = 0; i < this.components.Length; i++)
             {
-                return -1;
+                this.components[i].DcPredictor = 0;
             }
 
-            // Convert the huffman code to the symbol id
-            int c = (int)(((this.codeBuffer >> (32 - k)) & Bmask[k]) + table.ValOffset[k]);
-
-            // Convert the id to a symbol
-            this.codeBits -= k;
-            this.codeBuffer <<= k;
-            return table.Values[c];
+            this.eobrun = 0;
+            this.scanBuffer.Reset();
         }
 
         [MethodImpl(InliningOptions.ShortMethod)]
-        private int ExtendReceive(int n)
+        private bool HandleRestart()
         {
-            if (this.codeBits < n)
+            if (this.restartInterval > 0 && (--this.todo) == 0)
             {
-                this.FillBuffer();
+                this.todo = this.restartInterval;
+
+                if (this.scanBuffer.HasRestart())
+                {
+                    this.Reset();
+                    return true;
+                }
+
+                if (this.scanBuffer.Marker != JpegConstants.Markers.XFF)
+                {
+                    this.stream.Position = this.scanBuffer.MarkerPosition;
+                    this.Reset();
+                    return true;
+                }
             }
 
-            int sgn = (int)this.codeBuffer >> 31;
-            uint k = LRot(this.codeBuffer, n);
-            this.codeBuffer = k & ~Bmask[n];
-            k &= Bmask[n];
-            this.codeBits -= n;
-            return (int)(k + (Bias[n] & ~sgn));
+            return false;
         }
 
-        [MethodImpl(InliningOptions.ShortMethod)]
-        private void CheckBits()
+        internal sealed class HuffmanScanBuffer
```

Comment:
I suggest to turn this into a `struct` making short static methods of `HuffmanScanDecoder` (like `GetBits`) member methods `HuffmanScanBuffer`.
Of course, we can deal with this microoptimization opportunity in a follow-up PR.

Reply:
Great idea! I toyed with turning the class into a struct but didn't think to move the methods. The functionality is much better encapsulated now and reads much better!